src/main/scala/mesosphere/marathon/core/deployment/impl/DeploymentActor.scala

```diff
@@ -188,8 +192,9 @@ private class DeploymentActor(
     val instances = await(instanceTracker.specInstances(runSpec.id))
 
     logger.info(s"Killing all instances of ${runSpec.id}: ${instances.map(_.instanceId)}")
-    await(Future.sequence(instances.map(i => instanceTracker.setGoal(i.instanceId, Goal.Decommissioned))))
-    await(killService.killInstances(instances, KillReason.DeletingApp))
+    val instancesAreTerminal = killService.watchForKilledInstances(instances)
+    await(Future.sequence(instances.map(i => instanceTracker.setGoal(i.instanceId, Goal.Decommissioned, GoalChangeReason.DeletingApp))))
+    await(instancesAreTerminal)
```

Comment:
This pattern seems to be repeating itself and might be error prone since we have to declare the watch *before* we set the goal. What do you think about following up with something that allows
```scala
await(Future.sequence(instances.map(i => instanceTracker.setGoal(i.instanceId, Goal.Decommissioned, GoalChangeReason.DeletingApp))))
await(killService.allTerminal(instances)
```

Reply:
Yeah we need to improve on that for sure. I created MARATHON-8540 - I want to explore few ideas otherwise we can groom that ticket once this PR is merged

Reply:
Any reason we don't go with my proposed solution? https://github.com/mesosphere/marathon/pull/6680 Am I missing something?

Reply:
https://github.com/mesosphere/marathon/pull/6680 ?